[typing] mypy errors when trying to extract headers when type is `LooseHeaders`

### Describe the bug

When trying to read the `headers` from `aiohttp.ClientResponseError` mypy complains,

```py
error: No overload variant of "get" of "Mapping" matches argument type "str"  [call-overload]
note: Possible overload variants:
note:     def get(self, istr, /) -> str | None
note:     def [_T] get(self, istr, /, default: str | _T) -> str | _T
Found 1 error in 1 file (checked 1 source file)
```


### To Reproduce

<details>

<summary>test.py</summary>

```py
from collections.abc import Iterable, Mapping
from typing import Optional, Tuple, Union

from aiohttp.typedefs import LooseHeaders
from multidict import CIMultiDict, CIMultiDictProxy, istr


def get_my_header(headers: LooseHeaders) -> Optional[str]:
    if isinstance(headers, Mapping):
        return headers.get("foo")
    return next((v for k, v in headers if k == "foo"), None)
```

</details>

```
❯ mypy test.py
test.py:17: error: No overload variant of "get" of "Mapping" matches argument type "str"  [call-overload]
test.py:17: note: Possible overload variants:
test.py:17: note:     def get(self, istr, /) -> str | None
test.py:17: note:     def [_T] get(self, istr, /, default: str | _T) -> str | _T
Found 1 error in 1 file (checked 1 source file)
```

### Expected behavior

`headers.get()` should not this throw type error.

### Logs/tracebacks

```python-traceback
N/A
```


### Python Version

```console
$ python --version
Python 3.10.8
```


### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.10.3
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: .venv/lib/python3.10/site-packages
Requires: aiohappyeyeballs, aiosignal, async-timeout, attrs, frozenlist, multidict, yarl
Required-by: aiobotocore, s3fs
```


### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.4
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: .venv/lib/python3.10/site-packages
Requires: 
Required-by: aiohttp, yarl
```


### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.3
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: .venv/lib/python3.10/site-packages
Requires: idna, multidict
Required-by: aiohttp
```


### OS

macOS

### Related component

Client

### Additional context

I believe this got introduced as a part of [#8620](https://github.com/aio-libs/aiohttp/pull/8620/files#diff-c84f87a2adc085f40d7283c8603485da23412b3a2ac00ea2d75fe0b9a411e6da) when we split `Mapping[Union[str, istr], str]` into `Mapping[str, str]` and `Mapping[istr, str]`. Reverting it to the older form works fine,

<details open>

<summary>test.py</summary>

```py
from collections.abc import Iterable, Mapping
from typing import Optional, Tuple, Union

from aiohttp.typedefs import LooseHeaders
from multidict import CIMultiDict, CIMultiDictProxy, istr

LooseHeadersFixed = Union[
    Mapping[Union[str, istr], str],
    CIMultiDict[str],
    CIMultiDictProxy[str],
    Iterable[Tuple[Union[str, istr], str]],
]


def get_my_header(headers: LooseHeadersFixed) -> Optional[str]:
    if isinstance(headers, Mapping):
        return headers.get("foo")
    return next((v for k, v in headers if k == "foo"), None)
```
</details>

```py
❯ mypy test.py
Success: no issues found in 1 source file
```

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

--- a/CHANGES/8768.bugfix.rst
+++ b/CHANGES/8768.bugfix.rst
@@ -1,0 +1,1 @@
+Used more precise type for ``ClientResponseError.headers``, fixing some type errors when using them -- by :user:`Dreamorcerer`.

--- a/aiohttp/client_exceptions.py
+++ b/aiohttp/client_exceptions.py
@@ -6,7 +6,7 @@ from typing import TYPE_CHECKING, Optional, Tuple, Union
 from multidict import MultiMapping
 
 from .http_parser import RawResponseMessage
-from .typedefs import LooseHeaders, StrOrURL
+from .typedefs import StrOrURL
 
 try:
     import ssl

--- a/aiohttp/client_exceptions.py
+++ b/aiohttp/client_exceptions.py
@@ -3,6 +3,8 @@
 import asyncio
 from typing import TYPE_CHECKING, Optional, Tuple, Union
 
+from multidict import MultiMapping
+
 from .http_parser import RawResponseMessage
 from .typedefs import LooseHeaders, StrOrURL
 
@@ -69,7 +71,7 @@ class ClientResponseError(ClientError):
         *,
         status: Optional[int] = None,
         message: str = "",
-        headers: Optional[LooseHeaders] = None,
+        headers: Optional[MultiMapping[str]] = None,
     ) -> None:
         self.request_info = request_info
         if status is not None:

--- a/tests/test_client_exceptions.py
+++ b/tests/test_client_exceptions.py
@@ -43,7 +43,7 @@ class TestClientResponseError:
             history=(),
             status=400,
             message="Something wrong",
-            headers=CIMultiDict(),
+            headers=CIMultiDict(foo="bar"),
         )
         err.foo = "bar"  # type: ignore[attr-defined]
         for proto in range(pickle.HIGHEST_PROTOCOL + 1):
@@ -53,7 +53,8 @@ class TestClientResponseError:
             assert err2.history == ()
             assert err2.status == 400
             assert err2.message == "Something wrong"
-            assert err2.headers == {}
+            # Use headers.get() to verify static type is correct.
+            assert err2.headers.get("foo") == "bar"
             assert err2.foo == "bar"
 
     def test_repr(self) -> None:

--- a/tests/test_client_exceptions.py
+++ b/tests/test_client_exceptions.py
@@ -78,7 +78,7 @@ class TestClientResponseError:
             history=(),
             status=400,
             message="Something wrong",
-            headers=CIMultiDict,
+            headers=CIMultiDict(),
         )
         assert str(err) == ("400, message='Something wrong', url='http://example.com'")
 

--- a/tests/test_client_exceptions.py
+++ b/tests/test_client_exceptions.py
@@ -43,7 +43,7 @@ class TestClientResponseError:
             history=(),
             status=400,
             message="Something wrong",
-            headers={},
+            headers=CIMultiDict(),
         )
         err.foo = "bar"  # type: ignore[attr-defined]
         for proto in range(pickle.HIGHEST_PROTOCOL + 1):
@@ -65,7 +65,7 @@ class TestClientResponseError:
             history=(),
             status=400,
             message="Something wrong",
-            headers={},
+            headers=CIMultiDict(),
         )
         assert repr(err) == (
             "ClientResponseError(%r, (), status=400, "
@@ -78,7 +78,7 @@ class TestClientResponseError:
             history=(),
             status=400,
             message="Something wrong",
-            headers={},
+            headers=CIMultiDict,
         )
         assert str(err) == ("400, message='Something wrong', url='http://example.com'")
 

--- a/tests/test_client_exceptions.py
+++ b/tests/test_client_exceptions.py
@@ -70,7 +70,7 @@ class TestClientResponseError:
         )
         assert repr(err) == (
             "ClientResponseError(%r, (), status=400, "
-            "message='Something wrong', headers={})" % (self.request_info,)
+            "message='Something wrong', headers=<CIMultiDict()>)" % (self.request_info,)
         )
 
     def test_str(self) -> None:

